### ISSUE TITLE
Note editor freezes after Spell checker is switched on

### ISSUE THREAD

**Issue by michaelkn** [d1vanov/quentier#149]
The issue can be reproduced with the following steps:
1. Create a blank note
2. Input any word (for me it reproduces with Russian, probably because Quentier haven't found Russian dictionary for spell checker)
3. Enable spell checker
4. Press enter

After that one cannot input anything in the note editor and a crash will occur on quit from application.

Minidump and logs attached:
[minidump.tar.gz](https://github.com/d1vanov/quentier/files/1569472/minidump.tar.gz)

**Comment by d1vanov** [comment]
Thank you very much for providing all the necessary troubleshooting info!

Some notes about what I found looking at the log:

1. There's a bug in libquentier's `NoteEditorPrivate::applySpellCheck` method: it forms wrong JavaScript if the list of note's misspelled words is empty:

> 2017-12-18 23:40:27.669 MSK libquentier/src/note_editor/NoteEditor_p.cpp @ 7135 [Debug]: NoteEditorPrivate::setSpellcheck: enabled = true
> 2017-12-18 23:40:27.669 MSK libquentier/src/note_editor/NoteEditor_p.cpp @ 5035 [Debug]: NoteEditorPrivate::enableSpellCheck
> 2017-12-18 23:40:27.669 MSK libquentier/src/note_editor/NoteEditor_p.cpp @ 5059 [Debug]: NoteEditorPrivate::refreshMisSpelledWordsList
> 2017-12-18 23:40:27.669 MSK libquentier/src/enml/ENMLConverter_p.cpp @ 1337 [Debug]: ENMLConverterPrivate::noteContentToPlainText: <en-note><div></div></en-note>
> 2017-12-18 23:40:27.669 MSK libquentier/src/note_editor/NoteEditor_p.cpp @ 5127 [Debug]: NoteEditorPrivate::applySpellCheck: apply to selection = false
> 2017-12-18 23:40:27.669 MSK libquentier/src/note_editor/NoteEditor_p.cpp @ 5143 [Trace]: Script: if (window.hasOwnProperty('spellChecker')) { spellChecker.appl); }
> 2017-12-18 23:40:27.669 MSK libquentier/src/note_editor/JavaScriptInOrderExecutor.cpp @ 40 [Trace]: JavaScriptInOrderExecutor: appended new script, there are 1 to execute now
> 2017-12-18 23:40:27.669 MSK libquentier/src/note_editor/NoteEditor_p.cpp @ 5160 [Debug]: NoteEditorPrivate::enableDynamicSpellCheck
> 2017-12-18 23:40:27.669 MSK libquentier/src/note_editor/JavaScriptInOrderExecutor.cpp @ 40 [Trace]: JavaScriptInOrderExecutor: appended new script, there are 1 to execute now
> 2017-12-18 23:40:27.673 MSK libquentier/src/note_editor/NoteEditorPage.cpp @ 171 [Debug]: NoteEditorPage::javaScriptConsoleMessage, message: Uncaught SyntaxError: Unexpected token ), level = 2, line number: 1, sourceID =

```c++
5124 void NoteEditorPrivate::applySpellCheck(const bool applyToSelection)                                                    
5125 {                                                                                                                       
5126     QNDEBUG(QStringLiteral("NoteEditorPrivate::applySpellCheck: apply to selection = ")                                 
5127             << (applyToSelection ? QStringLiteral("true") : QStringLiteral("false")));                                  
5128                                                                                                                         
5129     QString javascript = QStringLiteral("if (window.hasOwnProperty('spellChecker')) { spellChecker.apply");             
5130                                                                                                                         
5131     if (applyToSelection) {                                                                                             
5132         javascript += QStringLiteral("ToSelection");                                                                    
5133     }                                                                                                                   
5134                                                                                                                         
5135     javascript += QStringLiteral("('");                                                                                 
5136     for(auto it = m_currentNoteMisSpelledWords.constBegin(), end = m_currentNoteMisSpelledWords.constEnd(); it != end; ++it) {
5137         javascript += *it;                                                                                              
5138         javascript += QStringLiteral("', '");                                                                           
5139     }                                                                                                                   
5140     javascript.chop(3);     // Remove trailing ", '";                                                                   
5141     javascript += QStringLiteral("); }");                                                                               
5142                                                                                                                         
5143     QNTRACE(QStringLiteral("Script: ") << javascript);                                                                  
5144                                                                                                                         
5145     GET_PAGE()                                                                                                          
5146     page->executeJavaScript(javascript, NoteEditorCallbackFunctor<QVariant>(this, &NoteEditorPrivate::onSpellCheckSetOrCleared));
5147 }
```
2. Something goes very wrong on attempt to dynamically catch the word you wrote:

> 2017-12-18 23:40:28.814 MSK libquentier/src/note_editor/NoteEditor_p.cpp @ 6656 [Debug]: NoteEditorPrivate::onSpellCheckerDynamicHelperUpdate: тест
> 2017-12-18 23:40:28.814 MSK libquentier/src/note_editor/SpellChecker_p.cpp @ 124 [Debug]: SpellCheckerPrivate::checkSpell: тест
> 2017-12-18 23:40:28.814 MSK libquentier/src/note_editor/NoteEditor_p.cpp @ 6684 [Trace]: Current note's misspelled words: тест
> 2017-12-18 23:40:28.814 MSK libquentier/src/note_editor/NoteEditor_p.cpp @ 5127 [Debug]: NoteEditorPrivate::applySpellCheck: apply to selection = true
> 2017-12-18 23:40:28.814 MSK libquentier/src/note_editor/NoteEditor_p.cpp @ 5143 [Trace]: Script: if (window.hasOwnProperty('spellChecker')) { spellChecker.applyToSelection('тест'); }
> 2017-12-18 23:40:28.814 MSK libquentier/src/note_editor/JavaScriptInOrderExecutor.cpp @ 40 [Trace]: JavaScriptInOrderExecutor: appended new script, there are 1 to execute now
> <...>
> 2017-12-18 23:40:28.851 MSK libquentier/src/note_editor/NoteEditorPage.cpp @ 171 [Debug]: NoteEditorPage::javaScriptConsoleMessage, message: SpellChecker::applyToSelection, level = 0, line number: 50, sourceID =
> 2017-12-18 23:40:28.851 MSK libquentier/src/note_editor/NoteEditorPage.cpp @ 171 [Debug]: NoteEditorPage::javaScriptConsoleMessage, message: SpellChecker::buildRegex, level = 0, line number: 110, sourceID =
> 2017-12-18 23:40:28.851 MSK libquentier/src/note_editor/NoteEditorPage.cpp @ 171 [Debug]: NoteEditorPage::javaScriptConsoleMessage, message: Search regex: (?:^|[\b\s.,;:-=+-_><])()(?:[\b\s.,;:-=+-_><]|$); flags: gi, level = 0, line number: 196, sourceID =
> 2017-12-18 23:40:28.854 MSK libquentier/src/note_editor/NoteEditorPage.cpp @ 171 [Debug]: NoteEditorPage::javaScriptConsoleMessage, message: SpellChecker::applyToNode, level = 0, line number: 66, sourceID =
> 2017-12-18 23:40:28.857 MSK libquentier/src/note_editor/NoteEditorPage.cpp @ 171 [Debug]: NoteEditorPage::javaScriptConsoleMessage, message: Applying to node: value = null, text content =   .JColResizer{table-layout:fixed;} .JColResizer td, .JColResizer th{overflow:hidden;padding-left:0!important; padding-right:0!important;}  .JCLRgrips{ height:0px; position:relative;} .JCLRgrip{margin-left:-5px; position:absolute; z-index:5; } .JCLRgrip .JColResizer{position:absolute;background-color:red;filter:alpha(opacity=1);opacity:0;width:10px;height:100%;cursor: e-resize;top:0px} .JCLRLastGrip{position:absolute; width:1px; } .JCLRgripDrag{ border-left:1px dotted black;     } .JCLRFlex{width:auto!important;}тест, outer HTML = <html><head><meta http-equiv="Content-Type" content="text/html" chars
> et="UTF-8"><link rel="stylesheet" type="text/css" href="qrc:/css/jquery-ui.min.css"><link rel="stylesheet" type="text/css" href="qrc:/css/en-crypt.css"><link rel="stylesheet" type="text/css" href="qrc:/css/hover.css"><link rel="stylesheet" type="text/css" href="qrc:/css/en-decrypted.css"><link rel="stylesheet" type="text/css" href="qrc:/css/en-media-generic.css"><link rel="stylesheet" type="text/css" href="qrc:/css/en-media-image.css"><link rel="stylesheet" type="text/css" href="qrc:/css/image-area-hilitor.css"><link rel="stylesheet" type="text/css" href="qrc:/css/en-todo.css"><link rel="stylesheet" type="text/css" href="qrc:/css/link.css"><link rel="stylesheet" type="text/css" href="qrc:/css/misspell.css"><title></title><style type="text/css">  .JColResizer{table-layout:fixed;} .JColResizer td, .JColResizer th{overflow:hidden;padding-left:0!important; padding-right:0!important;}  .JCLRgrips{ height:0px; position:relative;} .JCLRgrip{margin-left:-5px; position:absolute; z-index:5; } .JCLRgrip .JColResizer{position:absolute;background-color:red;filter:alpha(opacity=1);opacity:0;width:10px;height:100%;cursor: e-resize;top:0px} .JCLRLastGrip{position:absolute; width:1px; } .JCLRgripDrag{ border-left:1px dotted black;   } .JCLRFlex{width:auto!important;}</style></head><body contenteditable="true">тест<div></div><div><br></div></body></html>, inner HTML = <head><meta http-equiv="Conte
> nt-Type" content="text/html" charset="UTF-8"><link rel="stylesheet" type="text/css" href="qrc:/css/jquery-ui.min.css"><link rel="stylesheet" type="text/css" href="qrc:/css/en-crypt.css"><link rel="stylesheet" type="text/css" href="qrc:/css/hover.css"><link rel="stylesheet" type="text/css" href="qrc:/css/en-decrypted.css"><link rel="stylesheet" type="text/css" href="qrc:/css/en-media-generic.css"><link rel="stylesheet" type="text/css" href="qrc:/css/en-media-image.css"><link rel="stylesheet" type="text/css" href="qrc:/css/image-area-hilitor.css"><link rel="stylesheet" type="text/css" href="qrc:/css/en-todo.css"><link rel="stylesheet" type="text/css" href="qrc:/css/link.css"><link rel="stylesh
> eet" type="text/css" href="qrc:/css/misspell.css"><title></title><style type="text/css">  .JColResizer{table-layout:fixed;} .JColResizer td, .JColResizer th{overflow:hidden;padding-left:0!important; padding-right:0!important;}  .JCLRgrips{ height:0px; position:relative;} .JCLRgrip{margin-left:-5px; position:absolute; z-index:5; } .JCLRgrip .JColResizer{position:absolute;background-color:red;filter:alpha(opacity=1);opacity:0;width:10px;height:100%;cursor: e-resize;top:0px} .JCLRLastGrip{position:absolute; width:1px; } .JCLRgripDrag{ border-left:1px dotted black; } .JCLRFlex{width:auto!important;}</style></head><body contenteditable="true">тест<div></div><div><br></div></body>, level = 0, line
>  number: 69, sourceID =
2017-12-18 23:40:28.905 MSK libquentier/src/note_editor/NoteEditorPage.cpp @ 171 [Debug]: NoteEditorPage::javaScriptConsoleMessage, message: SpellChecker::highlightMisSpelledWords, level = 0, line number: 201, sourceID =
> 2017-12-18 23:40:28.907 MSK libquentier/src/note_editor/NoteEditorPage.cpp @ 171 [Debug]: NoteEditorPage::javaScriptConsoleMessage, message: Found misspelled word(s) within the text:   .JColResizer{table-layout:fixed;} .JColResizer td, .JColResizer th{overflow:hidden;padding-left:0!important; padding-right:0!important;}  .JCLRgrips{ height:0px; position:relative;} .JCLRgrip{margin-left:-5px; position:absolute; z-index:5; } .JCLRgrip .JColResizer{position:absolute;background-color:red;filter:alpha(opacity=1);opacity:0;width:10px;height:100%;cursor: e-resize;top:0px} .JCLRLastGrip{position:absolute; width:1px; } .JCLRgripDrag{ border-left:1px dotted black;  } .JCLRFlex{width:auto!important;}; regs.length = 2, regs[0] =  , regs[1] = , regs.index = 0, matchRegex.lastIndex = 1, level =
>  0, line number: 258, sourceID =                                                                                        
> 2017-12-18 23:40:28.910 MSK libquentier/src/note_editor/NoteEditorPage.cpp @ 171 [Debug]: NoteEditorPage::javaScriptConsoleMessage, message: Found target text node for match: name = #text, node value:   .JColResizer{table-layout:fixed;} .JColResizer td, .JColResizer th{overflow:hidden;padding-left:0!important; padding-right:0!important;}  .JCLRgrips{ height:0px; position:relative;} .JCLRgrip{margin-left:-5px; position:absolute; z-index:5; } .JCLRgrip .JColResizer{position:absolute;background-color:red;filter:alpha(opacity=1);opacity:0;width:10px;height:100%;cursor: e-resize;top:0px} .JCLRLastGrip{position:absolute; width:1px; } .JCLRgripDrag{ border-left:1px dotted black;        } .JCLRFlex{width:auto!important;}; adjusted match index = 1, level = 0, line number: 291, sourceID =
> 2017-12-18 23:40:28.910 MSK libquentier/src/note_editor/NoteEditorPage.cpp @ 171 [Debug]: NoteEditorPage::javaScriptConsoleMessage, message: after: node value =  .JColResizer{table-layout:fixed;} .JColResizer td, .JColResizer th{overflow:hidden;padding-left:0!important; padding-right:0!important;}  .JCLRgrips{ height:0px; position:relative;} .JCLRgrip{margin-left:-5px; position:absolute; z-index:5; } .JCLRgrip .JColResizer{position:absolute;background-color:red;filter:alpha(opacity=1);opacity:0;width:10px;height:100%;cursor: e-resize;top:0px} .JCLRLastGrip{position:absolute; width:1px; } .JCLRgripDrag{ border-left:1px dotted black; } .JCLRFlex{width:auto!important;}, level = 0, line number: 323, sourceID =
> 2017-12-18 23:40:28.910 MSK libquentier/src/note_editor/NoteEditorPage.cpp @ 171 [Debug]: NoteEditorPage::javaScriptConsoleMessage, message: after + substring 0: node value =  .JColResizer{table-layout:fixed;} .JColResizer td, .JColResizer th{overflow:hidden;padding-left:0!important; padding-right:0!important;}  .JCLRgrips{ height:0px; position:relative;} .JCLRgrip{margin-left:-5px; position:absolute; z-index:5; } .JCLRgrip .JColResizer{position:absolute;background-color:red;filter:alpha(opacity=1);opacity:0;width:10px;height:100%;cursor: e-resize;top:0px} .JCLRLastGrip{position:absolute; width:1px; } .JCLRgripDrag{ border-left:1px dotted black;   } .JCLRFlex{width:auto!important;}, level = 0, line number: 332, sourceID =
> 2017-12-18 23:40:28.913 MSK libquentier/src/note_editor/NoteEditorPage.cpp @ 171 [Debug]: NoteEditorPage::javaScriptConsoleMessage, message: After misspelled word highlighting: inner HTML:  <em class="misspell"></em> .JColResizer{table-layout:fixed;} .JColResizer td, .JColResizer th{overflow:hidden;padding-left:0!important; padding-right:0!important;}  .JCLRgrips{ height:0px; position:relative;} .JCLRgrip{margin-left:-5px; position:absolute; z-index:5; } .JCLRgrip .JColResizer{position:absolute;background-color:red;filter:alpha(opacity=1);opacity:0;width:10px;height:100%;cursor: e-resize;top:0px} .JCLRLastGrip{position:absolute; width:1px; } .JCLRgripDrag{ border-left:1px dotted black;  } .JCLRFlex{width:auto!important;}, level = 0, line number: 368, sourceID =

I.e. here it inserted empty block for misspelled word highlighting: `<em class="misspell"></em>`. It is a bug in its own right. But it goes on:

> 2017-12-18 23:40:28.925 MSK libquentier/src/note_editor/NoteEditorPage.cpp @ 171 [Debug]: NoteEditorPage::javaScriptConsoleMessage, message: After misspelled word highlighting: inner HTML:  <em class="misspell"></em><em class="misspell"></em> .JColResizer{table-layout:fixed;} .JColResizer td, .JColResizer th{overflow:hidden;padding-left:0!important; padding-right:0!important;}  .JCLRgrips{ height:0px; position:relative;} .JCLRgrip{margin-left:-5px; position:absolute; z-index:5; } .JCLRgrip .JColResizer{position:absolute;background-color:red;filter:alpha(opacity=1);opacity:0;width:10px;height:100%;cursor: e-resize;top:0px} .JCLRLastGrip{position:absolute; width:1px; } .JCLRgripDrag{ border-left:1px dotted black;        } .JCLRFlex{width:auto!important;}, level = 0, line number: 368, sour
> ceID =
> <...>
> 2017-12-18 23:40:31.084 MSK libquentier/src/note_editor/NoteEditorPage.cpp @ 171 [Debug]: NoteEditorPage::javaScriptConsoleMessage, message: After misspelled word highlighting: inner HTML:  <em class="misspell"></em><em class="misspell"></em><em class="misspell"></em><em class="misspell"></em><em class="misspell"></em><em class="misspell"></em><em class="misspell"></em><em class="misspell"></em><em class="misspell"></em><em class="misspell"></em><em class="misspell"></em><em class="misspell"></em><em class="misspell"></em><em class="misspell"></em><em class="misspell"></em><em class="misspell"></em><em class="misspell"></em><em class="misspell"></em><em class="misspell"></em><em class="misspell"></em><em class="misspell"></em><em class="misspell"></em><em class="misspell"></em><e
m class="misspell"></em><em class="misspell"></em><em class="misspell"></em><em class="misspell"></em> .JColResizer{table-layout:fixed;} .JColResizer td, .JColResizer th{overflow:hidden;padding-left:0!important; padding-right:0!important;}  .JCLRgrips{ height:0px; position:relative;} .JCLRgrip{margin-left:-5px; position:absolute; z-index:5; } .JCLRgrip .JColResizer{position:absolute;background-color:red;filter:alpha(opacity=1);opacity:0;width:10px;height:100%;cursor: e-resize;top:0px} .JCLRLastGrip{position:absolute; width:1px; } .JCLRgripDrag{ border-left:1px dotted black;    } .JCLRFlex{width:auto!important;}, level = 0, line number: 368, sourceID =

So the bug's origins are more or less clear to me, they all belong to libquentier's note editor component, the JavaScript trickery attempting to highlight words considered misspelled.

**Comment by d1vanov** [comment]
It should be solved after commits [086927b](https://github.com/d1vanov/libquentier/commit/086927b9b91bf51a44a01b352d23dea0947f814f) and [56bfc5a](https://github.com/d1vanov/libquentier/commit/56bfc5aed00be0d0e96b7babb3d35a59406ba994) to libquentier in [master](https://github.com/d1vanov/libquentier/commit/master) and [development](https://github.com/d1vanov/libquentier/commit/development) branches.

I've found one other bug related to the spell checker - d1vanov/libquentier#39. But at least it shouldn't cause crashes.

**Comment by d1vanov** [comment]
I'll trigger the rebuilds on Travis CI and AppVeyor CI when CI gets finished for libquentier itself.
